### PR TITLE
Improvement about ComponentsProviderGenerator

### LINKED ISSUE #1742: Adding @Gauge annotated method in ApplicationScoped bean causing error in bean instantation
When I tried to add `@Gauge` metric to my app, it failed to start with 

```
Caused by: java.lang.NoSuchMethodError: WorkerProvider_Bean: method <init>()V not found
	at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Unknown Source)
	at io.quarkus.arc.ArcContainerImpl.<init>(ArcContainerImpl.java:95)
	at io.quarkus.arc.Arc.initialize(Arc.java:34)
	at io.quarkus.arc.runtime.ArcDeploymentTemplate.getContainer(ArcDeploymentTemplate.java:43)
	at io.quarkus.deployment.steps.ArcAnnotationProcessor$build5.deploy(Unknown Source)
````

But this class definetly have a no-args constructor

When I remove `@Gauge` annotated method application starts without any issue

### PR DESCRIPTION
**Pull request by jmartisk**
`ComponentsProviderGenerator` was not enhancing beans that have only their constructor intercepted.

Fixes #1742 

https://github.com/quarkusio/quarkus/issues/1742
@mkouba please review

### CODE REVIEW

**Review by mkouba — CHANGES_REQUESTED** [review]
Looks good! But it would be nice to have a test in `arc/tests`.

Also from perf POV we could probably optimize `BeanInfo.getBoundInterceptors()` a little bit. For example return `Collections.emptyList()` immediately if both `lifecycleInterceptors` and `interceptedMethods` are empty, or reduce the stream usage.

**Comment by jmartisk** [comment]
Done, I refactored the `getBoundInterceptors` to avoid streams, and added a simple test

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java [hunk 1]
```diff
@@ -31,6 +31,7 @@
 import java.util.Set;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
```
Comment (gsmet) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java]: Looks like this is not needed anymore?

Comment (jmartisk) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java]: Oh.. right

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java [hunk 2]
```diff
@@ -300,6 +300,33 @@ boolean hasDefaultDestroy() {
         return bound.isEmpty() ? Collections.emptyList() : bound.stream().distinct().sorted().collect(Collectors.toList());
     }
 
+    /**
+     *
+     * @return an ordered list of all interceptors associated with the bean
+     */
+    List<InterceptorInfo> getBoundInterceptors2() {
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java]: `getBoundInterceptors2()`? I think we should replace the original implementation and use this one in `ComponentsProviderGenerator` instead...

Comment (jmartisk) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java]: Aaaaaa damn me, I committed wrong code. Sorry.

Comment (jmartisk) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java]: Should be OK now, I hope. I will have to watch  more what I'm committing :)

**Review by mkouba — APPROVED** [review]

**Pull request merged**